letsencrypt/client/network2.py

```diff
@@ -0,0 +1,480 @@
+        except requests.exceptions.RequestException as error:
+            raise errors.NetworkError(error)
+        self._check_response(response, content_type=content_type)
+        return response
+
+    def _post(self, uri, data, content_type=JSON_CONTENT_TYPE, **kwargs):
+        """Send POST data.
+
+        :param str content_type: Expected ``Content-Type``, fails if not set.
+
+        :raises letsencrypt.acme.messages2.NetworkError:
+
+        :returns: HTTP Response
+        :rtype: `requests.Response`
+
+        """
+        logging.debug('Sending POST data: %s', data)
+        try:
+            response = requests.post(uri, data=data, **kwargs)
+        except requests.exceptions.RequestException as error:
+            raise errors.NetworkError(error)
+        logging.debug('Received response %s: %s', response, response.text)
+
+        self._check_response(response, content_type=content_type)
+        return response
+
+    @classmethod
+    def _regr_from_response(cls, response, uri=None, new_authzr_uri=None,
+                            terms_of_service=None):
+        terms_of_service = (
+            response.links['terms-of-service']['url']
+            if 'terms-of-service' in response.links else terms_of_service)
+
+        if new_authzr_uri is None:
+            try:
+                new_authzr_uri = response.links['next']['url']
+            except KeyError:
+                raise errors.NetworkError('"next" link missing')
+
+        return messages2.RegistrationResource(
+            body=messages2.Registration.from_json(response.json()),
+            uri=response.headers.get('Location', uri),
+            new_authzr_uri=new_authzr_uri,
+            terms_of_service=terms_of_service)
+
+    def register(self, contact=messages2.Registration._fields[
+            'contact'].default):
+        """Register.
+
+        :param contact: Contact list, as accpeted by `.RegistrationResource`
+        :type contact: `tuple`
+
+        :returns: Registration Resource.
+        :rtype: `.RegistrationResource`
+
+        :raises letsencrypt.client.errors.UnexpectedUpdate:
+
+        """
+        new_reg = messages2.Registration(contact=contact)
+
+        response = self._post(self.new_reg_uri, self._wrap_in_jws(new_reg))
+        assert response.status_code == httplib.CREATED  # TODO: handle errors
+
+        regr = self._regr_from_response(response)
+        if regr.body.key != self.key.public() or regr.body.contact != contact:
+            raise errors.UnexpectedUpdate(regr)
+
+        return regr
+
+    def update_registration(self, regr):
+        """Update registration.
+
+        :pram regr: Registration Resource.
+        :type regr: `.RegistrationResource`
+
+        :returns: Updated Registration Resource.
+        :rtype: `.RegistrationResource`
+
+        """
+        response = self._post(regr.uri, self._wrap_in_jws(regr.body))
+
+        # TODO: Boulder returns httplib.ACCEPTED
```

Comment:
I think httplib.ACCEPTED (202) might be OK for updating a registration, it could take time.  I tried to find the place in the spec because it is related to the issue https://github.com/letsencrypt/acme-spec/issues/89 but I was unsuccessful. Updating a registration isn't listed in the table and I was unable to find it in the paragraphs describing the update : /  Perhaps it should be filed as another issue in the spec... :(